src/trie.js

```diff
@@ -0,0 +1,97 @@
+ * of states of the form:
+ * ```js
+ * {
+ *   'scopes': [..],      // scopes to add when entering this state
+ *   'a': { ... },        // next state if the current character is 'a'
+ *   ...
+ *   'end': { ... },      // state if this is the end of the input
+ * ```
+ *
+ * This takes into account both scope expansion (kleene star in the input
+ * to the trie) and role expansion (kleene star in the pattern).  The
+ * resulting trie does *not* take into account recursive application of
+ * patterns.
+ *
+ * Note that, as applied to roles, the PATTERN for every rule begins with
+ * `assume:`, but nothing in this file is specific to that prefix.
+ */
+exports.generateTrie = (rules) => {
+  // generate the state for rules i .. n-1 starting at character position k.
+  // This assumes that the rules have the same value at positions < k, as
+  // assured by the sort order.  Thus we have three segments to work with,
+  // in order:
+  //         k
+  // i -> ...*   -- (A) terminal * in pattern
+  //      ...    -- (B) end of pattern (where pattern[k-1] != '*')
+  //      ...*a  \
+  //      ...a   -- (C) nonterminal * or other characters in pattern
+  //      ...ab  /
+  // n ->
+  let gen = (i, n, k) => {
+    let state = {};
+
+    let j = i;
+    while (j < n) {
+      // find the current segment, at indices [seg .. j-1]
+      let seg = j;
+      let current = rules[j].pattern[k]; // current character
+      if (current === '*' && rules[j].pattern.length === k + 1) {
+        // segment type A, terminal * in pattern
+        do {
+          j++;
+        } while (j < n && rules[j].pattern[k] == '*' && rules[j].pattern.length === k + 1);
+        state.scopes = _.flatten(rules.slice(seg, j).map(r => r.scopes));
+      } else if (current === undefined) {
+        // segment type B, end of pattern (with pattern[k-1] != '*')
+        do {
+          j++;
+        } while (j < n && rules[j].pattern.length === k);
+        state['end'] = {scopes: _.flatten(rules.slice(seg, j).map(r => r.scopes))};
+      } else {
+        // segment type C, nonterminal * or other characters
+        do {
+          j++;
+        } while (j < n && rules[j].pattern[k] === current);
+        state[current] = gen(seg, j, k+1);
+      }
+    }
+
+    // a terminal star in the input here (state['*']['end']) will match all
+    // patterns
+    let allScopes = _.flatten(rules.slice(i, n).map(r => r.scopes));
+    if (state['*']) {
+      state['*']['end'] = {scopes: allScopes};
+    } else {
+      state['*'] = {end: {scopes: allScopes}};
+    }
+
+    return state;
+  };
+
+  rules.sort((a, b) => scopeCompare(a.pattern, b.pattern));
+  return gen(0, rules.length, 0);
+};
+
+/**
+ * Executes the trie for a single input, returning the accumulated scopes.
+ */
+exports.executeTrie = (trie, input) => {
+  let state = trie;
+  let result = [];
+  let k = 0;
+
+  while (state) {
+    if (state.scopes) {
+      result = result.concat(state.scopes);
```

Comment:
If instead you did: `result.push({matchedAt: k, scopes: state.scopes})`
Then you won't need to have scopes containing `<..>` to be on the form `{index, scope}`.

Instead when you do `executeTrie` just:
```js
let parameterized = [];
executeTrie(trie, inputScope).map(({matchedAt, scopes}) => {
  if (matchedAt !== null) {
    parameterized = parameterized.concat(scopes.replace('<..>', inputScope.substr(matchedAt)));
  } else {
    parameterized = parameterized.concat(scopes);
  }
});
```

Reply:
I rather like the cleanliness of the trie implementation here.  This change requires that the trie implementation know something about parameterization *and* that the caller know something about parameterization.

I don't think what you've suggested would be noticeably more efficient, would it?  If not, I'd like to keep the cleaner separation.

Reply:
Ah, I see that this is useful for other purposes, too -- in particular, being able to sort and normalize `scopes` in the trie.

Reply:
Although it would have to be `{scopes: [{matchedAt, scope}]}` since each can match at a different position in the input.

Reply:
And indeed, the same scope may appear several times with different matchedAt:

```
x* --> b<..>c
xy* --> b<..>c
```
`executeTrie(true, 'xyz') -> `[{scope: `b<..>c', matchedAt: 1}, {scope: 'b<..>c', matchedAt: 2}]`.

Reply:
I should stop writing comments until I figure this out.